materials/property_test.py

```diff
@@ -0,0 +1,176 @@
+        prop = StateDependentProperty('name', yaml_dict)
+
+        # Verification
+        self.assertEqual(2.0, prop.default_value)
+        self.assertEqual('MPa', prop.units)
+        self.assertEqual('mmpds', prop.reference)
+        self.assertEqual(len(prop.state_vars), 1)
+        self.assertEqual(prop.state_vars[0], 'temperature')
+
+    def test_query_1d(self):
+        """Test query_value with a 1-d lookup table."""
+        # Setup
+        dv = 2.0
+        yaml_dict = {
+            'default_value': dv,
+            'units': 'MPa',
+            'reference': 'mmpds',
+            'variation_with_state': {
+                'state_vars': ['temperature'],
+                'state_vars_units': ['kelvin'],
+                'value_type': 'multiplier',
+                'representation': 'table',
+                'temperature': np.arange(4),
+                'values': np.arange(4)**2,
+            }}
+        prop = StateDependentProperty('name', yaml_dict)
+
+        # Action and verification
+        # Check queries at given points
+        self.assertEqual(prop.query_value({'temperature': 1}), 1**2 * dv)
+        self.assertEqual(prop.query_value({'temperature': 2}), 2**2 * dv)
+        # Check interpolation between given points
+        self.assertAlmostEqual(prop.query_value({'temperature': 2.5}, method='linear'), 6.5 * dv)
+        # Check bad query
+        with self.assertRaises(ValueError):
+            prop.query_value({'fish': 1.})    # fish is not a state variable.
+
+    def test_init_2d(self):
+        """Test init with a 2-d lookup table."""
+        # Setup
+        yaml_dict = {
+            'default_value': 2.0,
+            'units': 'MPa',
+            'reference': 'mmpds',
+            'variation_with_state': {
+                'state_vars': ['exposure time', 'temperature'],
+                'state_vars_units': ['hour', 'kelvin'],
+                'value_type': 'multiplier',
+                'representation': 'table',
+                'exposure time': {
+                    0.0: {
+                        'temperature': np.arange(4),
+                        'values': np.arange(4)**2
+                    },
+                    0.1: {
+                        'temperature': np.arange(4),
+                        'values': (np.arange(4) + 0.1)**2
+                    },
+                }
+            }}
+
+        # Action
+        prop = StateDependentProperty('name', yaml_dict)
+
+        # Verification
+        self.assertEqual(2.0, prop.default_value)
+        self.assertEqual('MPa', prop.units)
+        self.assertEqual('mmpds', prop.reference)
+        self.assertEqual(len(prop.state_vars), 2)
+        self.assertEqual(prop.state_vars[0], 'exposure time')
+        self.assertEqual(prop.state_vars[1], 'temperature')
+
+        # Verification of "private" attributes
+        # These tests are tied to the implementation, not interface
+        # Is this sketchy? Should the test only depend on the interface?
```

Comment:
Yes, it's noble if the test only tests the promises made by the interface; but I'll put aside zealotry for a moment

Here's a soft rubric I use:
- Try to rigorously test the contracts in the public interface
- If that doesn't exercise a private subcomponent `X` that you'd like to be confident in, break `X` into a separate thing with a public interface and a contract
- Sometimes that's onerous and doesn't buy you much. Then testing `X` with the original interface is fine, just make sure the user knows they're not allowed to assume things about `X`.


Another note:
The people who say "only test your public interface" are often assuming a very intensive test battery

Reply:
alleviated by https://github.com/mvernacc/material-properties-interchange/pull/1/commits/f48f19c3a749890accddc1232ecf53498025fc36